aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -98,6 +101,46 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
   return at::arange(like.size(dim), like.options().dtype(at::kLong));
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~ complex / complex_polar ~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor& complex_out(Tensor& result, const Tensor& real, const Tensor& imag) {
+  // Sort of hacky, but necessary so that input dtypes don't get promoted to complex.
```

Comment:
Making these a comparison op is VERY hacky and should not be done.

Just construct the appropriate TensorIteratorConfig yourself. Also, type promotion in TensorIterator now only considers input, not output, tensors, so having a complex output won't promote the input tensors to complex internally.

Reply:
You might want to just verify that real and image have the same dtype, that dtype is float32 or float64, and that out has the correct corresponding complex type here.

Reply:
Sounds good. About the dtypes, I originally did that but @anjali411 suggested I have type promotion. Should I switch back? I'll add the out check.

Reply:
Adding type promotion wouldn't be hard but I'd suggest you simplify your life and we worry about that later (and separately) if it comes up.

Reply:
It's already implemented in this PR.

Reply:
Oh I think I see what you mean, I don't have it implemented for `complex_out`. It seems the use of a separate `complex_out` here doesn't add much value while being tedious for checks. Does every PyTorch function need to support the out argument? If not, would an exception be suitable here?

Reply:
It's typical to support `out` although you're right that the value added is often small (`out` is not popular). I think it's OK to not support an `out` argument to start, although I don't know how much easier that makes things.

Reply:
I see. What I'm thinking right now is that `complex` just uses `real`s dtype and the function for converting from float to complex (maybe the function can throw an error if `real`'s dtype isn't floating) to construct `result`'s dtype. Then `complex_out` will check same dtype, that `real` is floating, and then that `result` is the corresponding complex dtype via the same function from earlier. Any thoughts?

Reply:
That sounds correct.

Reply:
Cool. I'll try to find if the function's already implemented.

Reply:
It is:

https://github.com/pytorch/pytorch/blob/e6221f4ca16e7d9fa9179db625d8dea7ea6af02b/c10/core/DefaultDtype.cpp#L20

Reply:
Oh wait, you need a slightly different function, sorry. @anjali411 can tell you if there's a function mapping from float type to corresponding complex type.

Reply:
I'm a bit confused, should I be calling `set_default_dtype(real.scalar_type())` and then `get_default_complex_dtype()`?

Reply:
No no no: absolutely not. Sorry, like I said, I was mistaken about the function you need. You do want a function mapping the given float dtypes of the input tensors to the corresponding complex type.

Reply:
Ah didn't see your second comment, got it.